# Capstone project scraping

In [48]:
import re
import requests
from bs4 import BeautifulSoup

import pandas as pd
import numpy as np

import geocoder

## Function to get entire listing

In [49]:
def get_listings(result):
    
    listings = []
    
    for d in result.findAll(class_ = "placard-details module"):

        listings.append(d)

    return listings

## Scraping first page only

In [59]:
# Loopnet banned my ip address while scrapping, so had to use a proxy IP address

proxies = {
    'http':'http://192.25.162.203'
    }

url = 'http://www.loopnet.com/for-lease/new-york-ny/retail/'

r = requests.get(url, proxies=proxies)

page = r.content
soup = BeautifulSoup(page, 'lxml')

results = (get_listings(soup))

In [60]:
# Displying the results of the scrapping

results

[<div class="placard-details module">\n<h5 class="listing-address">\n<a class="column-09" href="http://www.loopnet.com/Listing/19621797/134-West-37th-Street-New-York-NY/" ng-click="RedirectToProfile('http://www.loopnet.com/Listing/19621797/134-West-37th-Street-New-York-NY/'); $event.stopPropagation();" title="134 West 37th Street">\n<address itemprop="address" itemscope="" itemtype="http://schema.org/PostalAddress">\n<span class="street-address" itemprop="streetAddress" title="134 West 37th Street">134 West 37th Street</span>\n<span itemprop="addressLocality" title="New York, NY">New York</span>,\r\n    <span itemprop="addressState" title="NY, NY">NY</span>\n<span itemprop="postalCode" title="10018">10018</span>\n</address>\n</a>\n<span class="listing-price column-04">$45-$79.20 SF/Year</span>\n</h5>\n<ul>\n<li>\n<span>Office Bldg</span>\n</li>\n<li>\n<i>2,500 - 10,000</i>\n<span>SF Available</span>\n</li>\n<li>\n<i>2</i>\n<span>Spaces</span>\n</li>\n<li>\n<i>120,000</i>\n<span>SF Bldg

Every listing should include a list of :

1. listing webpage
2. street address, locality, state, postalcode
3. listing price
4. listing type (office building, restaurant, etc)
5. SF available
6. No. of spaces
7. SF building
8. property description

In [186]:
# Functions to get listing details from scrapped website(s)

def get_listing_details(listing_placard):
    
    listing = []
    
    # listing website
    try:
        listing.append(re.findall('href="(.+)" ng-', \
                                  str(listing_placard))[0])
        
         
    except:
        listing.append('')
    
    # listing streetaddress
    try:
        listing.append(re.findall('streetAddress" title="(.+)"', \
                                  str(listing_placard))[0])
    except:
        listing.append('')
     
    # listing address locality
    try:
        listing.append(re.findall('addressLocality" title="(.+)"', \
                                  str(listing_placard))[0])
    except:
        listing.append('')
    
    # listing state
    try:
        listing.append(re.findall('addressState" title="(.+),', \
                                  str(listing_placard))[0])
    except:
        listing.append('')
    
    # listing postcode
    try:
        listing.append(re.findall('postalCode" title="(\d+)"', \
                                  str(listing_placard))[0])
    except:
        listing.append('')
    
    # listing price
    try:
        listing.append(re.findall('listing-price.+>(.+)<', \
                                  str(listing_placard))[0])
    except:
        listing.append('')
    
    # listing type
    try:
        listing.append(re.findall('h5>\n<ul>\n<li>\n<span>(.+)<', \
                                  str(listing_placard))[0])
    except:
        listing.append('')
    
    # listing square feet available
    try:
        listing.append(re.findall('>(.+)</i>\n<span>SF Available', \
                                  str(listing_placard))[0])
    except:
        listing.append('')

    # listing spaces available
    try:
        listing.append(re.findall('>(.+)</i>\n<span>Spaces', \
                                  str(listing_placard))[0])
    except:
        listing.append('')
    
    # listing building square feet
    try:
        listing.append(re.findall('>(.+)</i>\n<span>SF Bldg', \
                                  str(listing_placard))[0])
    except:
        listing.append('')
    
    #listing property description
    try:
        listing.append(re.findall('property-description">((.|\n|\r)*)<', \
                                  str(listing_placard))[0])
    except:
        listing.append('')
    
    return listing

In [62]:
# Appending all the relevant listing details to an element within page_list
# Every element in the list is then a row of a single listing

page_list = []
for i in results:
    page_list.append(get_listing_details(i))

In [63]:
# Nameing the columns
col = ['Web_address', 'Street_address', 'Locality', 'State', 'Postcode', \
       'Price', 'Property_type', 'SF_avail', 'Spaces_avail', 'Bldg_SF', \
       'Prop_desc']

# Converting data into a dataframe
df = pd.DataFrame(page_list, columns= col)
df

Web_address          Street_address  \
0   http://www.loopnet.com/Listing/19621797/134-We...    134 West 37th Street   
1   http://www.loopnet.com/Listing/19852777/517-Co...     517 Columbus Avenue   
2   http://www.loopnet.com/Listing/19584842/76-Woo...           76 Wooster St   
3   http://www.loopnet.com/Listing/19632482/117-Ea...    117 East 15th Street   
4   http://www.loopnet.com/Listing/19872387/250-Hu...           250 Hudson St   
5   http://www.loopnet.com/Listing/19872550/485-Ma...         485 Madison Ave   
6   http://www.loopnet.com/Listing/18803035/315-Ma...         315 Madison Ave   
7   http://www.loopnet.com/Listing/19872441/315-Hu...           315 Hudson St   
8   http://www.loopnet.com/Listing/20002133/261-Fi...           261 Fifth Ave   
9   http://www.loopnet.com/Listing/19271844/100-Br...            100 Broad St   
10  http://www.loopnet.com/Listing/19963876/143-Or...          143 Orchard St   
11  http://www.loopnet.com/Listing/18922237/142-14...  142-144 west 26 street   
12  http://www.loopnet.com/Listing/19814141/382-La...    382 Lafayette Street   
13  http://www.loopnet.com/Listing/19857624/155-We...    155 West 18th Street   
14  http://www.loopnet.com/Listing/20056601/155-Ch...         155 Chambers St   
15  http://www.loopnet.com/Listing/20056597/70-74-...      70-74 Lafayette St   
16  http://www.loopnet.com/Listing/20054534/735-Ni...           735 Ninth Ave   
17  http://www.loopnet.com/Listing/20051678/274-Gr...            274 Grand St   
18  http://www.loopnet.com/Listing/20051336/580-59...      580-590 W 161st St   
19  http://www.loopnet.com/Listing/20050911/1143-1...            1143 1ST Ave   

        Locality State Postcode                 Price Property_type  \
0   New York, NY    NY    10018    $45-$79.20 SF/Year   Office Bldg   
1   New York, NY    NY    10024       $204.60 SF/Year    Restaurant   
2   New York, NY    NY    10012       $80-$84 SF/Year                 
3   New York, NY    NY    10003  $110.67-$120 SF/Year                 
4   New York, NY    NY    10013       $75-$90 SF/Year   Office Bldg   
5   New York, NY    NY    10022      $54-$550 SF/Year   Office Bldg   
6   New York, NY    NY    10017      $60-$355 SF/Year   Office Bldg   
7   New York, NY    NY    10013     $100-$175 SF/Year   Office Bldg   
8   New York, NY    NY    10001                  Rate                 
9   New York, NY    NY    10004           $94 SF/Year                 
10  New York, NY    NY    10002          $180 SF/Year  Other Retail   
11  New York, NY    NY    10001           $85 SF/Year                 
12  New York, NY    NY    10003       $218.18 SF/Year                 
13  New York, NY    NY    10011                  Rate                 
14  New York, NY    NY    10007          $216 SF/Year                 
15  New York, NY    NY    11209        $58.50 SF/Year                 
16  New York, NY    NY    10019                  Rate                 
17  New York, NY    NY    10002                  Rate                 
18  New York, NY    NY    10032                  Rate                 
19  New York, NY    NY    10065       $187.20 SF/Year                 

          SF_avail Spaces_avail  Bldg_SF  \
0   2,500 - 10,000            2  120,000   
1            2,346                 2,345   
2    3,000 - 5,000            2   10,000   
3    3,300 - 6,600            3    6,600   
4   3,100 - 15,832            3  394,424   
5     654 - 31,472            3  302,000   
6    1,802 - 3,894            4  148,370   
7    3,922 - 5,881            5  485,577   
8   1,590 - 21,457            3  444,716   
9            8,000                10,000   
10             800                 6,472   
11           4,000                53,054   
12           2,200                13,000   
13   2,440 - 7,406            3    7,406   
14             750                 8,900   
15           4,000                16,900   
16           2,000                 2,700   
17   1,514 - 3,664            2    5,944   
18          

In [64]:
# Checking to see if we got all 20 listings on the page
len(page_list)

# End of single page scrap testing

20

## Starting to scrap Loopnet.com for all their NYC listings

In [189]:
# Engineering the web address to account for page numbers
url_template1 = "http://www.loopnet.com/for-lease/ny/retail/"
url_template2 = "/?view=list"

# Maximum available pages on the website
max_pages = 25

# List of possible proxy IP addresses
"""58.176.46.248
4.31.142.200
64.20.45.138"""

# Proxy IP address

proxies = {
    'http':'http://4.31.142.200'
    }

# Starting with page 2 since we have already scrapped first page 
# from the test above

# Added "+1" since python range doesn't include upper limit
for pg in range(2, max_pages+1, 1):
    
    # Engineering the web address for each page
    url = url_template1 + str(pg) + url_template2

    r = requests.get(url, proxies=proxies)
    if str(r) == "<Response [200]>":

        page = r.content
        soup = BeautifulSoup(page, 'lxml')

        results.append(get_listings(soup))
        for i in results:
            page_list.append(get_listing_details(i))

    else:
        print "This url is not responding favorably."
        print "url =", url
        print "response =", r

In [190]:
# Converting data into dataframe

col = ['Web_address', 'Street_address', 'Locality', 'State', \
       'Postcode', 'Price', 'Property_type', 'SF_avail', \
       'Spaces_avail', 'Bldg_SF', 'Prop_desc']

df = pd.DataFrame(page_list, columns= col)

In [191]:
# Checking to see number of listings scraped
df.shape

(1238, 11)

In [192]:
# Saving to csv!
df.to_csv('listings.csv')

In [199]:
# Dropping any duplicates
df1 = df.drop_duplicates()

In [200]:
# Checking to see how many listings remained
df1.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 42 entries, 0 to 892
Data columns (total 14 columns):
Web_address            42 non-null object
Street_address         42 non-null object
Locality               42 non-null object
State                  42 non-null object
Postcode               42 non-null object
Price                  33 non-null object
Property_type          42 non-null object
SF_avail               42 non-null object
Spaces_avail           42 non-null object
Bldg_SF                42 non-null object
Prop_desc              42 non-null object
Full_street_address    42 non-null object
Latitude               20 non-null float64
Longitude              20 non-null float64
dtypes: float64(2), object(12)
memory usage: 4.9+ KB


### Comments:

Loopnet.com seemed to have many repeated entries. From an original universe set of 1,238 listings, there were only 42 entries left. This made me suspicious of the quality of data from Loopnet.com. Hence, this dataset was dropped for other sources.

# Scraping from CityFeet. using Selenium and PhantomJS

In [153]:
from selenium import webdriver
from selenium.webdriver.common.desired_capabilities import DesiredCapabilities
import time

In [269]:
# CityFeet.com using AJAX (Asynchronous JavaScript and XML), so Selenium
# with PhantomJS were used to automatically scroll to the bottom

pause = 5

driver = webdriver.PhantomJS(
    executable_path = \
    '/Users/peidacai/Downloads/phantomjs-2.1.1-macosx/bin/phantomjs')

dcap = dict(DesiredCapabilities.PHANTOMJS)

dcap['phantomjs.page.settings.userAgent'] = \
("Mozilla/5.0 (Macintosh; Intel Mac OS X 10_11_6) \
AppleWebKit/537.36 (KHTML, like Gecko) Chrome/54.0.2840.71 Safari/537.36")

driver = webdriver.PhantomJS(
    executable_path='/Users/peidacai/Downloads/\
    phantomjs-2.1.1-macosx/bin/phantomjs', \
    desired_capabilities = dcap, service_args = [
        '--ignore-ssl-errors=true'])

driver.implicitly_wait(20)
driver.set_window_size(1024, 768)
driver.get('http://www.cityfeet.com/cont/new-york-retail-space#')

# Code to automatically wait 5 seconds for new listings to load before 
# scrolling further. Website stopped loading if this is not included

lastHeight = driver.execute_script("return document.body.scrollHeight")
while True:
    driver.execute_script("window.scrollTo(0, document.body.scrollHeight);")
    time.sleep(pause)
    newHeight = driver.execute_script("return document.body.scrollHeight")
    if newHeight == lastHeight:
        break
    lastHeight = newHeight

## Appending html messages to list

In [270]:
soup = BeautifulSoup(driver.page_source, "html.parser")

results = soup.find_all('div','property clearfix src-CF level-30')

cityfeet_results =[]
for d in results:
    cityfeet_results.append(d)


In [271]:
len(cityfeet_results)

12

In [272]:
results = soup.find_all('div','property clearfix src-LN level-20')

for d in results:
    cityfeet_results.append(d)

In [273]:
len(cityfeet_results)

1774

In [274]:
results = soup.find_all('div','property clearfix src-LN level-10')

for d in results:
    cityfeet_results.append(d)

In [275]:
len(cityfeet_results)

1774

In [276]:
results = soup.find_all('div','property clearfix src-LN level-40')

for d in results:
    cityfeet_results.append(d)

In [277]:
len(cityfeet_results)

1777

## Defining function to sieve out listing details

In [183]:
def get_cityfeet_details(listing_placard):
    
    listing = []
    
    # listing streetaddress
    try:
        listing.append(re.findall('title="(.+)">', str(listing_placard))[0])
    except:
        listing.append('')
         
    # listing state
    try:
        listing.append((re.findall('<address>\n<div>(.+)</div>', \
                                   str(listing_placard))[0]).strip())
    except:
        listing.append('')
    
    # listing price
    try:
        listing.append(re.findall('"price">(.+)</strong><strong>', \
                                  str(listing_placard))[0])
    except:
        listing.append('')
        
    # listing square feet available
    try:
        listing.append(re.findall('<strong>(.+)</strong>', \
                                  str(listing_placard))[0])
    except:
        listing.append('')
        
    #listing property description
    try:
        listing.append(re.findall(
                '"propertydescription"><p>((.|\n|\r)*)</p>', \
                str(listing_placard))[0])
    except:
        listing.append('')
    
    return listing

In [278]:
# Adding listing details to dataframe

cityfeet_list = []

for i in cityfeet_results:
    cityfeet_list.append(get_cityfeet_details(i))

dfc = pd.DataFrame(cityfeet_list, columns = ['Street_address', 'State', \
                                             'Price', 'SF_avail', \
                                             'Prop_desc'])

In [279]:
dfc.head()

Street_address               State  \
0                           327 Guy Lombardo Avenue  Freeport, NY 11520   
1  327 W 57th St. PRIME RESTAURANT-RETAIL FOR LEASE  New York, NY 10019   
2                       118 West 14th Street Retail  New York, NY 10011   
3              Prime Retail, 5th Avenue Ladies Mile  New York, NY 10011   
4          BELOW MARKET!! PRIME MEATPACKING 14th ST  New York, NY 10014   

              Price        SF_avail  \
0    Rent Upon Req.  Size Upon Req.   
1  $118.75 /sf/year        3,840 SF   
2  $140.00 /sf/year        4,200 SF   
3    Rent Upon Req.        2,500 SF   
4   $65.00 /sf/year        3,100 SF   

                                           Prop_desc  
0  (Busy Intersection on corner of Guy Lombardo a...  
1  (Amazing Location! Perfect for Restaurant or R...  
2  (Prime retail at the base of a luxury resident...  
3  (This impeccably maintained boutique cast iron...  
4  (Boutique building located in the heart of the...

# Initial data cleaning

## Cleaning "Price" column

In [387]:
# Removing listings without prices
df_c = dfc[dfc['Price']!="Rent Upon Req."]

### Comments:
Lost about 600 entries, from 1,777 to 1,134.

In [389]:
df_c[df_c['Price'].str.contains("/sf/year")].shape

(1134, 5)

### Comments:
All prices are in the same unit, "square feet per year". So we can leave that alone for now.

## Cleaning address

### Zipcodes

In [390]:
# Extracting zipcodes from address
def get_zips(x):
    return re.findall('\d{5}', x)[0]

In [391]:
df_c.loc[:,'Postcode'] = df_c['State'].apply(get_zips)

In [392]:
df_c.head()

Street_address  \
1   327 W 57th St. PRIME RESTAURANT-RETAIL FOR LEASE   
2                        118 West 14th Street Retail   
4           BELOW MARKET!! PRIME MEATPACKING 14th ST   
10                                  778 Nostrand Ave   
14                                    5 Court Square   

                         State             Price  SF_avail  \
1           New York, NY 10019  $118.75 /sf/year  3,840 SF   
2           New York, NY 10011  $140.00 /sf/year  4,200 SF   
4           New York, NY 10014   $65.00 /sf/year  3,100 SF   
10          Brooklyn, NY 11215   $39.54 /sf/year  1,669 SF   
14  Long Island City, NY 11101   $35.00 /sf/year  3,500 SF   

                                            Prop_desc Postcode  
1   (Amazing Location! Perfect for Restaurant or R...    10019  
2   (Prime retail at the base of a luxury resident...    10011  
4   (Boutique building located in the heart of the...    10014  
10  (Prime Retail Space located on the main strip ...    11215  
14  (Immediate possession,  2   5 year lease term,...    11101

### Street address

In [393]:
# Function to create a boolean mask to weed out non-conventional addresses
def get_proper_add(x):
    
    # List of 'proper' address endings, to weed out erroneous ones
    proper_add = ['street', 'avenue', 'st', 'st.', 'ave', 'ave.', \
                  'broadway', 'place', 'pl.', 'pl',\
                  'square', 'cresent', 'road', 'highway', 'turnpike', \
                  'boulevard', 'steinway', '9', 'bypass', 'court', \
                  'parkway', 'lane', 'drive', 'americas', 'bowery', 'plaza']
    
    # Changing to all lower cases so it is easier to compare 
    x = x.lower()
    # 
    x = x.split()
    
    # Addresses ending with words in "proper_add" and digits are ok
    if (x[-1] in proper_add) or (re.match('\d+', x[-1])):
        return False
    else:
        return True

In [394]:
# Adding a boolean mask column
df_c['Address_to_correct'] = df_c['Street_address'].apply(get_proper_add)

/anaconda/lib/python2.7/site-packages/ipykernel/__main__.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  if __name__ == '__main__':


In [395]:
# Applying the filter to the dataframe
df_d = df_c[df_c['Address_to_correct'] == True]
df_d.shape

(26, 7)

In [396]:
# Inspecting the addresses
df_d

Street_address  \
1     327 W 57th St. PRIME RESTAURANT-RETAIL FOR LEASE   
2                          118 West 14th Street Retail   
26                                  17 Knolls Crescent   
181                                      4205 Avenue D   
375                                      2913 Avenue W   
427                                      1314 Avenue U   
548                                585 Veterans Road W   
593                                        95 Avenue B   
657                           25 Gardenville Parkway W   
834                                      1224 Avenue J   
848                                        33 Chestnut   
915                       1025 Little East Neck Road N   
1006                              2855 Veterans Road W   
1030                              237 Vestal Parkway E   
1044                         485 Schutt Road Extension   
1059                                  130 7th Avenue S   
1066                                  421 447 Avenue P   
1087                              2103 Grand Concourse   
1243                  230 Oriskany Boulevard Yorkville   
1336                   1395 Madison Avenue approximate   
1356                         505 Schutt Road Extension   
1357                         505 Schutt Road Extension   
1462                             3179 Erie Boulevard E   
1519                                     1211 Avenue J   
1670                                405 Central Park W   
1765                                 328 Canal Address   

                        State                    Price            SF_avail  \
1          New York, NY 10019         $118.75 /sf/year            3,840 SF   
2          New York, NY 10011         $140.00 /sf/year            4,200 SF   
26            Bronx, NY 10463          $53.04 /sf/year              905 SF   
181        Brooklyn, NY 11203          $39.00 /sf/year    1,700 - 3,400 SF   
375        Brooklyn, NY 11229          $33.00 /sf/year            2,200 SF   
427        Brooklyn, NY 11229          $49.00 /sf/year            1,200 SF   
548   Staten Island, NY 10309          $40.00 /sf/year    1,000 - 2,400 SF   
593        New York, NY 10009          $89.61 /sf/year            1,540 SF   
657         Buffalo, NY 14224           $8.00 /sf/year            3,532 SF   
834        Brooklyn, NY 11230          $90.00 /sf/year            3,300 SF   
848     Cold Spring, NY 10516          $36.00 /sf/year            2,000 SF   
915    West Babylon, NY 11704          $20.00 /sf/year    1,000 - 3,500 SF   
1006  Staten Island, NY 10309          $55.00 /sf/year    2,200 - 8,544 SF   
1030         Vestal, NY 13850           $8.00 /sf/year  10,000 - 12,240 SF   
1044     Middletown, NY 10940          $15.00 /sf/year            2,500 SF   
1059       New York, NY 10014         $225.00 /sf/year            2,000 SF   
1066       Brooklyn, NY 11223          $50.00 /sf/year              720 SF   
1087          Bronx, NY 10453          $78.22 /sf/year              675 SF   
1243      Yorkville, NY 13495  $9.00 - $12.00 /sf/year      400 - 3,500 SF   
1336      Manhattan, NY 10029         $124.29 /sf/year              700 SF   
1356     Middletown, NY 10940          $16.00 /sf/year            3,500 SF   
1357     Middletown, NY 10940          $18.00 /sf/year            1,500 SF   
1462       Syracuse, NY 13214          $17.00 /sf/year            1,745 SF   
1519       Brooklyn, NY 11230          $60.00 /sf/year            2,000 SF   
1670       New York, NY 10025         $116.40 /sf/year              750 SF   
1765       New York, NY 10013          $80.00 /sf/year            5,000 SF   

                                              Prop_desc Postcode  \
1     (Amazing Location! Perfect for Restaurant or R...    10019   
2     (Prime retail at the base of a luxury resident...    10011   
26                                      (Strip Mall, l)    10463   
181   (Coldwell Banker Reliable Commercial Division ...    11203   
375   (This is a brand new

In [397]:
# Correcting individual addresses

df_c.loc[1, 'Street_address'] = '327 W 57th St'
df_c.loc[2, 'Street_address'] = '118 West 14th Street'
df_c.loc[214, 'Street_address'] = '26 W 14th Street'
df_c.loc[1336, 'Street_address'] = '1395 Madison Avenue'
df_c.loc[1765, 'Street_address'] = '328 Canal Street'
df_c.loc[1356, 'Street_address'] = '505 Schutt Road'
df_c.loc[1357, 'Street_address'] = '505 Schutt Road'
df_c.loc[1044, 'Street_address'] = '485 Schutt Road'

/anaconda/lib/python2.7/site-packages/ipykernel/__main__.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy


In [398]:
df_c.head()

Street_address                       State  \
1                              327 W 57th St          New York, NY 10019   
2                       118 West 14th Street          New York, NY 10011   
4   BELOW MARKET!! PRIME MEATPACKING 14th ST          New York, NY 10014   
10                          778 Nostrand Ave          Brooklyn, NY 11215   
14                            5 Court Square  Long Island City, NY 11101   

               Price  SF_avail  \
1   $118.75 /sf/year  3,840 SF   
2   $140.00 /sf/year  4,200 SF   
4    $65.00 /sf/year  3,100 SF   
10   $39.54 /sf/year  1,669 SF   
14   $35.00 /sf/year  3,500 SF   

                                            Prop_desc Postcode  \
1   (Amazing Location! Perfect for Restaurant or R...    10019   
2   (Prime retail at the base of a luxury resident...    10011   
4   (Boutique building located in the heart of the...    10014   
10  (Prime Retail Space located on the main strip ...    11215   
14  (Immediate possession,  2   5 year lease term,...    11101   

   Address_to_correct  
1                True  
2                True  
4               False  
10              False  
14              False

In [401]:
df_c.shape

(1135, 7)

## Reverting to cleaning price column

In [403]:
# Function to get the digits in the price columns
def get_price(x):
    try:
        return re.findall('\d+\.\d\d', x)[0]
    except:
        return np.nan

In [404]:
# Applying function to the price column
df_c['Price_per_sf_per_year'] = df_c['Price'].apply(get_price)

In [406]:
# Sieving out rows with "NaN" price
df_c[df_c['Price_per_sf_per_year'].isnull()]

Street_address State Price SF_avail Prop_desc Postcode  \
214  26 W 14th Street   NaN   NaN      NaN       NaN      NaN   

    Address_to_correct Price_per_sf_per_year  
214                NaN                   NaN

In [317]:
df_c.head()

Street_address                       State  \
1                              327 W 57th St          New York, NY 10019   
2                       118 West 14th Street          New York, NY 10011   
4   BELOW MARKET!! PRIME MEATPACKING 14th ST          New York, NY 10014   
10                          778 Nostrand Ave          Brooklyn, NY 11215   
14                            5 Court Square  Long Island City, NY 11101   

               Price  SF_avail  \
1   $118.75 /sf/year  3,840 SF   
2   $140.00 /sf/year  4,200 SF   
4    $65.00 /sf/year  3,100 SF   
10   $39.54 /sf/year  1,669 SF   
14   $35.00 /sf/year  3,500 SF   

                                            Prop_desc Postcode  \
1   (Amazing Location! Perfect for Restaurant or R...    10019   
2   (Prime retail at the base of a luxury resident...    10011   
4   (Boutique building located in the heart of the...    10014   
10  (Prime Retail Space located on the main strip ...    11215   
14  (Immediate possession,  2   5 year lease term,...    11101   

   Address_to_correct Price_per_sf_per_year  
1                True                118.75  
2                True                140.00  
4               False                 65.00  
10              False                 39.54  
14              False                 35.00

In [407]:
# Dropping NaN row (row 214) and row with ambiguous address (row 4)
df_c.drop(4, axis = 0, inplace = True)
df_c.drop(214, axis = 0, inplace = True)
df_c.head()

Street_address                       State             Price  \
1          327 W 57th St          New York, NY 10019  $118.75 /sf/year   
2   118 West 14th Street          New York, NY 10011  $140.00 /sf/year   
10      778 Nostrand Ave          Brooklyn, NY 11215   $39.54 /sf/year   
14        5 Court Square  Long Island City, NY 11101   $35.00 /sf/year   
23       16 W 8th Street          New York, NY 10011  $149.61 /sf/year   

    SF_avail                                          Prop_desc Postcode  \
1   3,840 SF  (Amazing Location! Perfect for Restaurant or R...    10019   
2   4,200 SF  (Prime retail at the base of a luxury resident...    10011   
10  1,669 SF  (Prime Retail Space located on the main strip ...    11215   
14  3,500 SF  (Immediate possession,  2   5 year lease term,...    11101   
23  1,163 SF  (Excellent space steps from Washington Square ...    10011   

   Address_to_correct Price_per_sf_per_year  
1                True                118.75  
2                True                140.00  
10              False                 39.54  
14              False                 35.00  
23              False                149.61

In [408]:
# Confirming there are no more NaN in price column
df_c[df_c['Price_per_sf_per_year'].isnull()]

Empty DataFrame
Columns: [Street_address, State, Price, SF_avail, Prop_desc, Postcode, Address_to_correct, Price_per_sf_per_year]
Index: []

## Extracting city

In [409]:
df_c.loc[:,'City'] = df_c["State"].apply(lambda x: (str(x)).split(',')[0])
df_c.head(20)

Street_address                       State  \
1             327 W 57th St          New York, NY 10019   
2      118 West 14th Street          New York, NY 10011   
10         778 Nostrand Ave          Brooklyn, NY 11215   
14           5 Court Square  Long Island City, NY 11101   
23          16 W 8th Street          New York, NY 10011   
24            3821 Broadway          New York, NY 10032   
26       17 Knolls Crescent             Bronx, NY 10463   
27       256 W 109th Street          New York, NY 10025   
28        270 E 10th Street          New York, NY 10009   
30   120 Christopher Street          New York, NY 10014   
31          15 W 8th Street          New York, NY 10011   
32            97 1st Avenue          New York, NY 10003   
33      203 Chrystie Street          New York, NY 10002   
34       114 Stanton Street          New York, NY 10002   
35        22 E Merrick Road          Freeport, NY 11520   
36  964 Middle Country Road       Saint James, NY 11780   
37       294 Waverly Avenue         Patchogue, NY 11772   
38     4760 Sunrise Highway           Bohemia, NY 11716   
39      92 E Suffolk Avenue     Central Islip, NY 11722   
40  1815 E Jericho Turnpike        Huntington, NY 11743   

                         Price           SF_avail  \
1             $118.75 /sf/year           3,840 SF   
2             $140.00 /sf/year           4,200 SF   
10             $39.54 /sf/year           1,669 SF   
14             $35.00 /sf/year           3,500 SF   
23            $149.61 /sf/year           1,163 SF   
24  $140.00 - $140.24 /sf/year       830 - 966 SF   
26             $53.04 /sf/year             905 SF   
27            $205.26 /sf/year             380 SF   
28             $65.16 /sf/year           1,200 SF   
30            $129.00 /sf/year           1,200 SF   
31             $95.01 /sf/year           2,000 SF   
32             $81.58 /sf/year           2,500 SF   
33  $112.50 - $120.00 /sf/year     700 - 1,600 SF   
34            $138.46 /sf/year           1,300 SF   
35             $14.00 /sf/year           6,400 SF   
36             $24.00 /sf/year           3,700 SF   
37             $17.14 /sf/year           1,400 SF   
38             $20.00 /sf/year  3,200 - 14,050 SF   
39             $16.00 /sf/year           1,800 SF   
40             $24.00 /sf/year           9,000 SF   

                                            Prop_desc Postcode  \
1   (Amazing Location! Perfect for Restaurant or R...    10019   
2   (Prime retail at the base of a luxury resident...    10011   
10  (Prime Retail Space located on the main strip ...    11215   
14  (Immediate possession,  2   5 year lease term,...    11101   
23  (Excellent space steps from Washington Square ...    10011   
24  (Great corner location in a busy area of Washi...    10032   
26                                    (Strip Mall, l)    10463   
27                       (one story street retail, l)    10025   
28  (Ready to operate,Please do not talk to any em...    10009   
30  (1200 sf  basement 400 sf   30 seats   Vietnam...    10014   
31  (Great location in west village close to fifth...    10011   
32  (Beautiful corner restaurant   2,500 ground fl...    10003   
33  (1600 sf space divisible or taken as is locate...    10002   
34  (114 Stanton street one of the most incredible...    10002   
35  (To Watch a Video of this property please go t...    11520   
36  (:  Multi level building for lease located at ...    11780   
37  (Retail/Office building for sale or lease loca...    11772   
38  (Prime location at the corner of Sunrise Highw...    11716   
39  (Includes bar, boarding house and rental room ...    11722   
40  (Retail space available on heavily traveled Je...    11743   

   Address_to_correct Price_per_sf_per_year              City  
1                True                118.75          New York  
2                True                140.00          New York  
10              False                 39.54          Brooklyn  
14              

### Defining lists with only New York City zipcodes

In [331]:
manhattan_zip = pd.read_csv('manhattan_zip.txt', sep='\n')

In [333]:
brooklyn_zip = pd.read_csv('brooklyn_zip.txt', sep = '\n')

In [334]:
bronx_zip = pd.read_csv('bronx.txt', sep = '\n')

In [335]:
queens_zip = pd.read_csv('queens.txt', sep = '\n')

In [336]:
statten_zip = pd.read_csv('statten.txt', sep ='\n')

In [364]:
statten_zip = statten_zip.T.values.tolist()[0]

In [366]:
manhattan_zip = manhattan_zip.T.values.tolist()[0]

In [367]:
bronx_zip = bronx_zip.T.values.tolist()[0]
queens_zip = queens_zip.T.values.tolist()[0]
brooklyn_zip = brooklyn_zip.T.values.tolist()[0]

In [380]:
# NYC zipcodes less Statten Island
nyc_less_stat_zip = manhattan_zip + bronx_zip + brooklyn_zip + queens_zip

In [381]:
# Saving to txt file for possible future uses
nyc_less_stat_zip = pd.Series(nyc_less_stat_zip)
nyc_less_stat_zip.to_csv('nyc_less_stat_zip.txt')

In [368]:
# All NYC zipcodes
nyc_zip = manhattan_zip + bronx_zip + brooklyn_zip + queens_zip + statten_zip
nyc_zip

[10001,
 10002,
 10003,
 10004,
 10005,
 10006,
 10007,
 10009,
 10010,
 10011,
 10012,
 10013,
 10014,
 10016,
 10017,
 10018,
 10019,
 10020,
 10021,
 10022,
 10023,
 10024,
 10025,
 10026,
 10027,
 10028,
 10029,
 10030,
 10031,
 10032,
 10033,
 10034,
 10035,
 10036,
 10037,
 10038,
 10039,
 10040,
 10044,
 10048,
 10065,
 10069,
 10075,
 10111,
 10115,
 10128,
 10280,
 10281,
 10282,
 10451,
 10452,
 10453,
 10454,
 10455,
 10456,
 10457,
 10458,
 10459,
 10460,
 10461,
 10462,
 10465,
 10466,
 10467,
 10468,
 10469,
 10470,
 10471,
 10472,
 10473,
 10474,
 10475,
 11201,
 11203,
 11204,
 11205,
 11206,
 11207,
 11208,
 11209,
 11210,
 11211,
 11212,
 11213,
 11214,
 11215,
 11216,
 11217,
 11218,
 11219,
 11220,
 11221,
 11222,
 11223,
 11224,
 11225,
 11226,
 11228,
 11229,
 11230,
 11231,
 11232,
 11233,
 11234,
 11235,
 11236,
 11237,
 11238,
 11239,
 11251,
 11001,
 11004,
 11005,
 11040,
 11096,
 11101,
 11102,
 11103,
 11104,
 11105,
 11106,
 11109,
 11354,
 11355,
 11356,


In [370]:
nyc_zip = pd.Series(nyc_zip)
nyc_zip.to_csv('nyc_zip.txt')

### Removing rows without zip code

In [410]:
df_c[df_c['Postcode'].isnull()]

Empty DataFrame
Columns: [Street_address, State, Price, SF_avail, Prop_desc, Postcode, Address_to_correct, Price_per_sf_per_year, City]
Index: []

### Checking number of entries in nyc

In [411]:
df_c['Postcode'] = df_c['Postcode'].astype(int)

In [412]:
df_c[df_c['Postcode'].isin(nyc_zip)].shape

(569, 9)

#### Comments:
569 entries in NYC seemed decent enough for modeling. However, due to the relative inaccessibility of Statten Island by taxi, I am considering droppping entries in Statten Island. Ideally, we would have sufficient listings in Manhattan to reduce variability in comparison, since taxi would feature less strongly as a mode of transport in other boroughs. However... (see below)

#### Number of entries in Manhattan

In [413]:
df_c[df_c['Postcode'].isin(manhattan_zip)].shape

(180, 9)

#### Comments:
There are only 180 entries in Manhattan, possibly less, after more cleaning.

#### Number of entries in NYC less Statten Island

In [414]:
df_c[df_c['Postcode'].isin(nyc_less_stat_zip)].shape

(554, 9)

#### Comments:
Thankfully, there were only a few listings in Statten Island, therefore, we can proceed to drop Statten Island listings.

## Cleaning "SF_avail" column

In [415]:
# Removing characters
df_c.loc[:, 'SF_avail'] = \
df_c['SF_avail'].apply(lambda x : x.replace(' SF', ''))

In [416]:
# Removing commas
df_c.loc[:, 'SF_avail'] = \
df_c['SF_avail'].apply(lambda x : x.replace(',', ''))

df_c.head()

Street_address                       State             Price  \
1          327 W 57th St          New York, NY 10019  $118.75 /sf/year   
2   118 West 14th Street          New York, NY 10011  $140.00 /sf/year   
10      778 Nostrand Ave          Brooklyn, NY 11215   $39.54 /sf/year   
14        5 Court Square  Long Island City, NY 11101   $35.00 /sf/year   
23       16 W 8th Street          New York, NY 10011  $149.61 /sf/year   

   SF_avail                                          Prop_desc  Postcode  \
1      3840  (Amazing Location! Perfect for Restaurant or R...     10019   
2      4200  (Prime retail at the base of a luxury resident...     10011   
10     1669  (Prime Retail Space located on the main strip ...     11215   
14     3500  (Immediate possession,  2   5 year lease term,...     11101   
23     1163  (Excellent space steps from Washington Square ...     10011   

   Address_to_correct Price_per_sf_per_year              City  
1                True                118.75          New York  
2                True                140.00          New York  
10              False                 39.54          Brooklyn  
14              False                 35.00  Long Island City  
23              False                149.61          New York

In [420]:
# Some entries had a range of values for listing floor size
# These would be multiple units for rent in the same address.
# Based on experience, listing would show the lowest price on the website
# as click-bait
# Therefore, only the largest floor area were extracted

def clean_sf(x):
    if type(x) == type(1):
        return int(x)
    else:
        # Returning only the larger floor area of the range 
        # (see comments above)
        return int(x.split('-')[1])

In [421]:
# Applying change to "SF_avail" column
df_cityfeet.loc[:, 'SF_avail'] = df_cityfeet['SF_avail'].apply(clean_sf)

## Forming a full street address (for geocoding)

In [422]:
df_c.head()

Street_address                       State             Price  \
1          327 W 57th St          New York, NY 10019  $118.75 /sf/year   
2   118 West 14th Street          New York, NY 10011  $140.00 /sf/year   
10      778 Nostrand Ave          Brooklyn, NY 11215   $39.54 /sf/year   
14        5 Court Square  Long Island City, NY 11101   $35.00 /sf/year   
23       16 W 8th Street          New York, NY 10011  $149.61 /sf/year   

   SF_avail                                          Prop_desc  Postcode  \
1      3840  (Amazing Location! Perfect for Restaurant or R...     10019   
2      4200  (Prime retail at the base of a luxury resident...     10011   
10     1669  (Prime Retail Space located on the main strip ...     11215   
14     3500  (Immediate possession,  2   5 year lease term,...     11101   
23     1163  (Excellent space steps from Washington Square ...     10011   

   Address_to_correct Price_per_sf_per_year              City  
1                True                118.75          New York  
2                True                140.00          New York  
10              False                 39.54          Brooklyn  
14              False                 35.00  Long Island City  
23              False                149.61          New York

In [431]:
# Converting zipcodes to str type for concatenation
df_c['zips'] = df_c['Postcode'].astype(str)

In [432]:
# Forming full street address
df_c['Full_Street_address'] = df_c['Street_address'] + ', ' + df_c['City'] + ', NY ' + df_c['zips']
df_c.head()

Street_address                       State             Price  \
1          327 W 57th St          New York, NY 10019  $118.75 /sf/year   
2   118 West 14th Street          New York, NY 10011  $140.00 /sf/year   
10      778 Nostrand Ave          Brooklyn, NY 11215   $39.54 /sf/year   
14        5 Court Square  Long Island City, NY 11101   $35.00 /sf/year   
23       16 W 8th Street          New York, NY 10011  $149.61 /sf/year   

   SF_avail                                          Prop_desc  Postcode  \
1      3840  (Amazing Location! Perfect for Restaurant or R...     10019   
2      4200  (Prime retail at the base of a luxury resident...     10011   
10     1669  (Prime Retail Space located on the main strip ...     11215   
14     3500  (Immediate possession,  2   5 year lease term,...     11101   
23     1163  (Excellent space steps from Washington Square ...     10011   

   Address_to_correct Price_per_sf_per_year              City  \
1                True                118.75          New York   
2                True                140.00          New York   
10              False                 39.54          Brooklyn   
14              False                 35.00  Long Island City   
23              False                149.61          New York   

                           Full_Street_address   zips  
1            327 W 57th St, New York, NY 10019  10019  
2     118 West 14th Street, New York, NY 10011  10011  
10        778 Nostrand Ave, Brooklyn, NY 11215  11215  
14  5 Court Square, Long Island City, NY 11101  11101  
23         16 W 8th Street, New York, NY 10011  10011

In [433]:
# Extracting only relevant columns of the dataframe

df_cityfeet = pd.concat([df_c['Full_Street_address'], \
                         df_c['SF_avail'], \
                         df_c['Price_per_sf_per_year'], \
                         df_c['Prop_desc']], \
                        axis = 1)
df_cityfeet.reset_index(inplace = True, drop = True)
df_cityfeet.head()

Full_Street_address SF_avail Price_per_sf_per_year  \
0           327 W 57th St, New York, NY 10019     3840                118.75   
1    118 West 14th Street, New York, NY 10011     4200                140.00   
2        778 Nostrand Ave, Brooklyn, NY 11215     1669                 39.54   
3  5 Court Square, Long Island City, NY 11101     3500                 35.00   
4         16 W 8th Street, New York, NY 10011     1163                149.61   

                                           Prop_desc  
0  (Amazing Location! Perfect for Restaurant or R...  
1  (Prime retail at the base of a luxury resident...  
2  (Prime Retail Space located on the main strip ...  
3  (Immediate possession,  2   5 year lease term,...  
4  (Excellent space steps from Washington Square ...

In [434]:
# Checking types
df_cityfeet.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1133 entries, 0 to 1132
Data columns (total 4 columns):
Full_Street_address      1133 non-null object
SF_avail                 1133 non-null object
Price_per_sf_per_year    1133 non-null object
Prop_desc                1133 non-null object
dtypes: object(4)
memory usage: 35.5+ KB


In [435]:
# Converting floor area to float type
df_cityfeet.loc[:, 'Price_per_sf_per_year'] = \
df_cityfeet['Price_per_sf_per_year'].astype(float)

### Adding latitude and longitude

In [437]:
# Geocoding address with latitude and longitude

lat = []
lon = []
for i in df_cityfeet['Full_Street_address']:
    
    try:
        latlon = geocoder.google(i)
        lat.append(latlon.latlng[0])
        lon.append(latlon.latlng[1])
    
    except:
        lat.append(np.nan)
        lon.append(np.nan)
    
df_cityfeet['Latitude'] = lat
df_cityfeet['Longitude'] = lon
df_cityfeet.head()   

Full_Street_address SF_avail  Price_per_sf_per_year  \
0           327 W 57th St, New York, NY 10019     3840                 118.75   
1    118 West 14th Street, New York, NY 10011     4200                 140.00   
2        778 Nostrand Ave, Brooklyn, NY 11215     1669                  39.54   
3  5 Court Square, Long Island City, NY 11101     3500                  35.00   
4         16 W 8th Street, New York, NY 10011     1163                 149.61   

                                           Prop_desc   Latitude  Longitude  
0  (Amazing Location! Perfect for Restaurant or R...  40.767498 -73.984111  
1  (Prime retail at the base of a luxury resident...  40.737629 -73.997937  
2  (Prime Retail Space located on the main strip ...  40.672089 -73.950522  
3  (Immediate possession,  2   5 year lease term,...  40.746630 -73.942833  
4  (Excellent space steps from Washington Square ...  40.732584 -73.997463

## Adding back zipcodes and city to facilitate EDA

In [448]:
city = df_c['City']

In [453]:
zipp = df_c['zips']
zipp

0       10019
1       10011
2       11215
3       11101
4       10011
5       10032
6       10463
7       10025
8       10009
9       10014
10      10011
11      10003
12      10002
13      10002
14      11520
15      11780
16      11772
17      11716
18      11722
19      11743
20      10547
21      11727
22      11211
23      11216
24      11218
25      11225
26      11226
27      11206
28      11206
29      11226
        ...  
1103    10075
1104    10022
1105    10573
1106    10013
1107    14217
1108    14224
1109    11385
1110    12590
1111    10541
1112    11040
1113    11434
1114    10562
1115    10598
1116    10549
1117    11426
1118    10570
1119    11222
1120    11557
1121    10012
1122    10519
1123    12524
1124    12603
1125    14850
1126    11706
1127    10013
1128    13203
1129    10549
1130    10467
1131    11101
1132    10024
Name: zips, dtype: object

In [454]:
# Adding "City" as a separate column, for possible EDA purposes later
df_cityfeet['City'] = city
df_cityfeet.head()

Full_Street_address SF_avail  Price_per_sf_per_year  \
0           327 W 57th St, New York, NY 10019     3840                 118.75   
1    118 West 14th Street, New York, NY 10011     4200                 140.00   
2        778 Nostrand Ave, Brooklyn, NY 11215     1669                  39.54   
3  5 Court Square, Long Island City, NY 11101     3500                  35.00   
4         16 W 8th Street, New York, NY 10011     1163                 149.61   

                                           Prop_desc   Latitude  Longitude  \
0  (Amazing Location! Perfect for Restaurant or R...  40.767498 -73.984111   
1  (Prime retail at the base of a luxury resident...  40.737629 -73.997937   
2  (Prime Retail Space located on the main strip ...  40.672089 -73.950522   
3  (Immediate possession,  2   5 year lease term,...  40.746630 -73.942833   
4  (Excellent space steps from Washington Square ...  40.732584 -73.997463   

               City  
0          New York  
1          New York  
2          Brooklyn  
3  Long Island City  
4          New York

In [455]:
# Adding zipcode column to facilitate EDA
df_cityfeet['Zipcode'] = zipp
df_cityfeet.head()

Full_Street_address SF_avail  Price_per_sf_per_year  \
0           327 W 57th St, New York, NY 10019     3840                 118.75   
1    118 West 14th Street, New York, NY 10011     4200                 140.00   
2        778 Nostrand Ave, Brooklyn, NY 11215     1669                  39.54   
3  5 Court Square, Long Island City, NY 11101     3500                  35.00   
4         16 W 8th Street, New York, NY 10011     1163                 149.61   

                                           Prop_desc   Latitude  Longitude  \
0  (Amazing Location! Perfect for Restaurant or R...  40.767498 -73.984111   
1  (Prime retail at the base of a luxury resident...  40.737629 -73.997937   
2  (Prime Retail Space located on the main strip ...  40.672089 -73.950522   
3  (Immediate possession,  2   5 year lease term,...  40.746630 -73.942833   
4  (Excellent space steps from Washington Square ...  40.732584 -73.997463   

               City Zipcode  
0          New York   10019  
1          New York   10011  
2          Brooklyn   11215  
3  Long Island City   11101  
4          New York   10011

## Basic EDA

In [257]:
df_cityfeet.describe()

SF_avail  Price_per_sf_per_year    Latitude   Longitude  \
count     131.000000             131.000000  131.000000  131.000000   
mean     8403.114504             134.322824   40.755279  -73.978563   
std     48950.738279              98.794998    0.038811    0.024065   
min       279.000000              18.000000   40.703561  -74.012465   
25%       800.000000              64.845000   40.724422  -74.000122   
50%      1500.000000             105.000000   40.738547  -73.986773   
75%      2694.000000             163.930000   40.785749  -73.953785   
max    520000.000000             600.000000   40.862353  -73.925560   

            Zipcode  
count    131.000000  
mean   10020.282443  
std       21.166110  
min    10001.000000  
25%    10010.000000  
50%    10014.000000  
75%    10028.500000  
max    10128.000000

In [258]:
# Checking for outliers (large commercial projects)
df_cityfeet[df_cityfeet['SF_avail'] > 50000]

Street_address  SF_avail  Price_per_sf_per_year  \
64         2340 5th Avenue, NY 10037    520000                   49.0   
67  2005 2023 Third Avenue, NY 10029    121500                   50.0   
68       315 E 94th Street, NY 10128    185000                   60.0   

                                            Prop_desc   Latitude  Longitude  \
64  (Build to Suit Commercial Development in Harle...  40.816982 -73.934953   
67  (Build to Suit Commercial Development in East ...  40.794119 -73.942374   
68  (Build to Suit Commercial Development in the U...  40.782479 -73.946914   

    Zipcode  
64    10037  
67    10029  
68    10128

In [458]:
# Checking to see eventual number of listings
df_cityfeet.shape

(1133, 8)

In [459]:
# Dropping duplicates
df_cityfeet = df_cityfeet.drop_duplicates()
df_cityfeet.shape

(1128, 8)

In [460]:
# Saving to csv
df_cityfeet.to_csv('cityfeet_retail_rental.csv')

In [461]:
pd.pivot_table(df_cityfeet, values = ['SF_avail', 'Price_per_sf_per_year'], \
               index = ['Zipcode'], aggfunc='count')

Price_per_sf_per_year  SF_avail
Zipcode                                 
10001                        3         3
10002                       21        21
10003                       11        11
10004                        1         1
10005                        1         1
10007                        3         3
10009                        4         4
10010                        1         1
10011                       14        14
10012                       12        12
10013                       13        13
10014                       16        16
10016                        1         1
10017                        3         3
10018                        3         3
10019                        2         2
10021                        3         3
10022                        6         6
10023                        1         1
10024                        5         5
10025                        4         4
10026                        3         3
10027                        4         4
10028                        2         2
10029                       11        11
10030                        1         1
10031                        3         3
10032                        4         4
10033                        6         6
10035                        3         3
...                        ...       ...
14437                        1         1
14445                        1         1
14450                        2         2
14456                        1         1
14464                        1         1
14472                        1         1
14482                        1         1
14519                        1         1
14526                        1         1
14543                        1         1
14559                        1         1
14568                        1         1
14569                        1         1
14580                        2         2
14604                        1         1
14606                        2         2
14607                        2         2
14611                        3         3
14612                        1         1
14614                        1         1
14615                        1         1
14617                        1         1
14618                        6         6
14622                        2         2
14623                        1         1
14624                        3         3
14625                        1         1
14626                        1         1
14701                        2         2
14850                        2         2

[394 rows x 2 columns]

In [267]:
pd.pivot_table(df_cityfeet, values = ['SF_avail', 'Price_per_sf_per_year'], \
               index = ['Zipcode'])

Price_per_sf_per_year       SF_avail
Zipcode                                      
10001               165.000000    2700.000000
10002               104.502667    1413.333333
10003               144.753000    1600.000000
10004                43.200000    1250.000000
10005                25.000000    2750.000000
10007               142.833333    1587.666667
10009                74.805000    1434.000000
10011               133.121250    3304.750000
10012               259.555455    2113.636364
10013               167.873750    2676.750000
10014               217.896000    1132.800000
10016                57.000000    5200.000000
10017                95.440000    1383.000000
10018                56.990000    7450.000000
10019               118.750000    3840.000000
10021               144.356667    2400.000000
10022               104.798000    1932.600000
10023               150.000000     800.000000
10024               208.842500    1442.750000
10025               150.205000    1330.000000
10026                64.890000    1165.000000
10027                82.996667    2486.666667
10028               200.820000    1050.000000
10029                57.577000   13801.300000
10030                60.000000    1560.000000
10031                78.500000     850.000000
10032                65.110000     853.333333
10033                79.592500    1240.000000
10035                68.055000    4850.000000
10036               155.000000    2350.000000
10037                49.000000  520000.000000
10038                50.000000    6000.000000
10040                80.000000    6000.000000
10065               124.425000     950.000000
10075               430.110000     279.000000
10128                73.793333   66166.666667

## Resources and references

1. Sources for property listings 
- http://www.loopnet.com/
- http://www.cityfeet.com/